### ISSUE TITLE
TypeError: Type parameter +RV without a default follows type parameter with a default in _inductor/utils.py

### ISSUE THREAD

**Issue by poedator** [pytorch/pytorch#140914]
### 🐛 Describe the bug

env: 
python 3.9, 
torch.__version__='2.5.1'  also observed in 2.4.1

when importing transformers and torch I get error triggered by typing problem in this line: 
`class CachedMethod(Protocol, Generic[P, RV]):`
https://github.com/pytorch/pytorch/blob/99014a297c179862af38ee86bac2051434d3db41/torch/_inductor/utils.py#L459C1-L459C46
`TypeError: Type parameter +RV without a default follows type parameter with a default`

suggegsted solution: 
replace  `class CachedMethod(Protocol, Generic[P, RV]):` with `class CachedMethod(Protocol, Generic[RV, P]):`
in https://github.com/pytorch/pytorch/blob/99014a297c179862af38ee86bac2051434d3db41/torch/_inductor/utils.py#L459C1-L459C46
possibly related to https://github.com/pytorch/pytorch/pull/127685 

ERROR STACK BELOW:
```
  File "/conda/envs/main/lib/python3.9/site-packages/transformers/utils/import_utils.py", line 1778, in _get_module
    return importlib.import_module("." + module_name, self.__name__)
  File "/conda/envs/main/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/conda/envs/main/lib/python3.9/site-packages/transformers/models/gpt2/modeling_gpt2.py", line 40, in <module>
    from ...modeling_utils import PreTrainedModel, SequenceSummary
  File "/conda/envs/main/lib/python3.9/site-packages/transformers/modeling_utils.py", line 48, in <module>
    from .loss.loss_utils import LOSS_MAPPING
  File "/conda/envs/main/lib/python3.9/site-packages/transformers/loss/loss_utils.py", line 19, in <module>
    from .loss_deformable_detr import DeformableDetrForObjectDetectionLoss, DeformableDetrForSegmentationLoss
  File "/conda/envs/main/lib/python3.9/site-packages/transformers/loss/loss_deformable_detr.py", line 4, in <module>
    from ..image_transforms import center_to_corners_format
  File "/conda/envs/main/lib/python3.9/site-packages/transformers/image_transforms.py", line 22, in <module>
    from .image_utils import (
  File "/conda/envs/main/lib/python3.9/site-packages/transformers/image_utils.py", line 58, in <module>
    from torchvision.transforms import InterpolationMode
  File "/conda/envs/main/lib/python3.9/site-packages/torchvision/__init__.py", line 10, in <module>
    from torchvision import _meta_registrations, datasets, io, models, ops, transforms, utils  # usort:skip
  File "/conda/envs/main/lib/python3.9/site-packages/torchvision/models/__init__.py", line 2, in <module>
    from .convnext import *
  File "/conda/envs/main/lib/python3.9/site-packages/torchvision/models/convnext.py", line 8, in <module>
    from ..ops.misc import Conv2dNormActivation, Permute
  File "/conda/envs/main/lib/python3.9/site-packages/torchvision/ops/__init__.py", line 23, in <module>
    from .poolers import MultiScaleRoIAlign
  File "/conda/envs/main/lib/python3.9/site-packages/torchvision/ops/poolers.py", line 10, in <module>
    from .roi_align import roi_align
  File "/conda/envs/main/lib/python3.9/site-packages/torchvision/ops/roi_align.py", line 7, in <module>
    from torch._dynamo.utils import is_compile_supported
  File "/conda/envs/main/lib/python3.9/site-packages/torch/_dynamo/__init__.py", line 39, in <module>
    from .polyfills import loader as _  # usort: skip # noqa: F401
  File "/conda/envs/main/lib/python3.9/site-packages/torch/_dynamo/polyfills/loader.py", line 22, in <module>
    POLYFILLED_MODULES: Tuple["ModuleType", ...] = tuple(
  File "/conda/envs/main/lib/python3.9/site-packages/torch/_dynamo/polyfills/loader.py", line 23, in <genexpr>
    importlib.import_module(f".{submodule}", package=polyfills.__name__)
  File "/conda/envs/main/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/conda/envs/main/lib/python3.9/site-packages/torch/_dynamo/polyfills/builtins.py", line 24, in <module>
    def all(iterable: Iterable[object], /) -> bool:
  File "/conda/envs/main/lib/python3.9/site-packages/torch/_dynamo/decorators.py", line 312, in wrapper
    rule_map: Dict[Any, Type[VariableTracker]] = get_torch_obj_rule_map()
  File "/conda/envs/main/lib/python3.9/site-packages/torch/_dynamo/trace_rules.py", line 2860, in get_torch_obj_rule_map
    obj = load_object(k)
  File "/conda/envs/main/lib/python3.9/site-packages/torch/_dynamo/trace_rules.py", line 2891, in load_object
    val = _load_obj_from_str(x[0])
  File "/conda/envs/main/lib/python3.9/site-packages/torch/_dynamo/trace_rules.py", line 2875, in _load_obj_from_str
    return getattr(importlib.import_module(module), obj_name)
  File "/conda/envs/main/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/conda/envs/main/lib/python3.9/site-packages/torch/_higher_order_ops/map.py", line 6, in <module>
    from torch._functorch.aot_autograd import AOTConfig, create_joint
  File "/conda/envs/main/lib/python3.9/site-packages/torch/_functorch/aot_autograd.py", line 20, in <module>
    from torch._inductor.utils import BoxedBool
  File "/conda/envs/main/lib/python3.9/site-packages/torch/_inductor/utils.py", line 456, in <module>
    class CachedMethod(Protocol, Generic[P, RV]):
  File "/conda/envs/main/lib/python3.9/typing.py", line 277, in inner
    return func(*args, **kwds)
  File "/conda/envs/main/lib/python3.9/typing.py", line 1005, in __class_getitem__
    return _GenericAlias(cls, params)
  File "/conda/envs/main/lib/python3.9/typing.py", line 746, in __init__
    self.__parameters__ = _collect_type_vars(params)
  File "/conda/envs/main/lib/python3.9/site-packages/yt/packages/typing_extensions.py", line 3019, in _collect_type_vars
    raise TypeError(f'Type parameter {t!r} without a default'
TypeError: Type parameter +RV without a default follows type parameter with a default

```

### Versions

python 3.9
torch.__version__='2.5.1'

cc @ezyang @malfet @xuzhao9 @gramster @chauhang @penguinwu

**Comment by fang-d** [comment]
I also met this problem, leading to a fatal error in my project. Hope it can be resolved soon 🤩.

**Comment by jlamperez** [comment]
Hi, I am also facing this problem. Any info?

**Comment by tanwarsh** [comment]
Hi, I'm blocked because of this issue. Hoping it will be resolved soon.

**Comment by shink** [comment]
Hi guys, I'd love to do this. Please let me know how to repro this issue, thanks!

BTW, it seems that this problem is caused by the order of parameters. Since this definition of passing parameters first and returning values ​​second is usually reasonable, the best fix is to update the parameter order in its caller, right?

**Comment by poedator** [comment]
> the best fix is to update the parameter order in its caller, right?

Right. It should be a simple 1-string PR with swapped typing arguments.

**Comment by kta-intel** [comment]
TL;DR - workaround monkey patch
```
import typing_extensions
sys.modules['pip._vendor.typing_extensions'] = typing_extensions
```
Replace `pip._vendor.typing_extensions` with the package your system is calling

Hoping for discussion around a more elegant solution though

------------

What seems to be happening for us (cc @tanwarsh) is a mismatch in `typing_extensions`. More specifically, it is a mismatch in the `NoDefault` class that is being checked in order to enforce default ordering (see: [typing_extensions.py#L3314](https://github.com/python/typing_extensions/blob/main/src/typing_extensions.py#L3314)).

`ParamSpec` in `torch._inductor.utils` is being imported from `typing_extensions` (see [L43](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/utils.py#L43)). 
Which is later set as:
```
P = ParamSpec("P")
RV = TypeVar("RV", covariant=True)

class CachedMethod(Protocol, Generic[P, RV]):
...
```
The order is actually fine from what I understand, since `P` actually isn't set with a default. The error arises when it is time to actually the parameters for default. For whatever reason, our codebase will actually use the `pip` vendored `typing_extensions` (i.e. `pip._vendor.typing_extensions` to check. What ends up happening is Param's default is set as `<class 'typing_extensions.NoDefaultType'>` whereas the NoDefault that is being compared is <class 'pip._vendor.typing_extensions.NoDefaultType'>. These two classes are not recognized as equivalent and so `has_default` gets set to True and `TypeError` gets thrown.

I was never able to create an exact reproducer using torch inductor's params, directly but I think this should be sufficient to get the point across:

```
from typing_extensions import ParamSpec
from pip._vendor.typing_extensions import NoDefault
from typing import (
    TypeVar,
)

P = ParamSpec("P")
RV = TypeVar("RV", covariant=True)

types = [P, RV]

for t in types:
    has_default = getattr(t, '__default__', NoDefault) is not NoDefault
    print(f"Type variable: {t}")
    default = getattr(t, '__default__', NoDefault)
    print(f"Type of default value: {type(default)}")
    print(f"Type of NoDefault singleton: {type(NoDefault)}")
    print(f"Has default value? {has_default}")

# Type variable: ~P
# Type of default value: <class 'typing_extensions.NoDefaultType'>
# Type of NoDefault singleton: <class 'pip._vendor.typing_extensions.NoDefaultType'>
# Has default value? True
# Type variable: +RV
# Type of default value: <class 'pip._vendor.typing_extensions.NoDefaultType'>
# Type of NoDefault singleton: <class 'pip._vendor.typing_extensions.NoDefaultType'>
# Has default value? False
```

You'll see that `P` gets marked as having a default value, but it doesn't.

@poedator 
I think you might be running into the same issue, based on this line from your error:
```
File "/conda/envs/main/lib/python3.9/site-packages/yt/packages/typing_extensions.py", line 3019, in _collect_type_vars
    raise TypeError(f'Type parameter {t!r} without a default'
TypeError: Type parameter +RV without a default follows type parameter with a default
```
I'd guess that your case in comparing `<class 'typing_extensions.NoDefaultType'>` versus `<class 'yt.packages..NoDefaultType'>`

I found that a workaround that works for us is to monkey patch `typing_extensions` to ensure it is the same module

i.e
```
import typing_extensions
sys.modules['pip._vendor.typing_extensions'] = typing_extensions
```
(Replace `pip._vendor.typing_extensions` with the package your system is calling)

@shink @malfet 
Swapping `P` and `RV` would probably work, but that is only because `RV` is a `TypeVar` imported from `typing`, so it is properly detected as having no default while ignoring `P`

Wonder if it is possible for import `ParamSpec` from typing instead of typing_extensions, or if there is a way to dynamically imported from whatever vendored typing_extension is detecting on system in an elegant way.

Guess we could also raise an issue upstream at typing_extensions to have a more robust way to detect "NoDefault" from different modules that'll hopefully get quickly adopted into the vendored versions?

I don't have an elegant solution at the top of my mind.

**Comment by tanwarsh** [comment]
@kta-intel , Thank you. workaround patch you suggested worked.

**Comment by poedator** [comment]
@kta-intel , thank you, the monkey patch works for me too.
but will there be a permanent solution on the torch side? would simple swapping typing arguments be enough?

**Comment by sychou** [comment]
Confirming that the bug still exists and patch still works with Python 3.12.8 and crewai 0.100.1.

**Comment by hauntsaninja** [comment]
What is `yt.packages.typing_extensions`? And what exactly is attempting to mess with `pip._vendor.typing_extensions`?

If I do e.g. `python -c 'import torch._dynamo'` in a fresh environment with latest torch and typing-extensions installed everything works fine. This suggests the issue isn't with pytorch or typing-extensions, but with some other cursed thing.

Is there a clear minimal repro? This will probably help figure out what is cursing y'all.

**Comment by tanwarsh** [comment]
@hauntsaninja, I couldn't reproduce the issue directly, but it occurs when using OpenFL while running the taskrunner workspace for torch/mnist (https://openfl.readthedocs.io/en/latest/tutorials/taskrunner.html#).

@poedator, I was able to resolve the issue with pip version 24.1.2 without the patch.

For details please refer https://github.com/securefederatedai/openfl/issues/1441.

**Comment by Bludator** [comment]
TLDR; don't use internal code or library that use it

---
I am using `micropipenv` which in turns imports pip like `from pip._internal.req import ...`.
This import, if used in *right* order, breaks the torch. So here is minimal repro for my case:

```python
import torch
import pip._vendor.typing_extensions
import torch._dynamo
```
The actual import that causes the problem could be any import of pip which internally uses the `pip._vendor.typing_extensions` which is like half of the files there.

---


I am still not sure what exactly is going on but I have found [pip._vendor.__init__](https://github.com/pypa/pip/blob/e00c5100e3c29828d980fa8e7acf4cf43ec875a9/src/pip/_vendor/__init__.py) is quite suspicious. Although, simply import of `pip._vendor` does not trigger the error it thinkers with the `sys.modules`.  


To add to @kta-intel :

Interestingly, all the types used in the `class CachedMethod(Protocol, Generic[P, RV]):` are actually correctly imported. 
In other words (= code 😉 ):

```python
print(ParamSpec.__module__, sys.modules[ParamSpec.__module__].__file__)
print(TypeVar.__module__, sys.modules[TypeVar.__module__].__file__)
print(Protocol.__module__, sys.modules[Protocol.__module__].__file__)
print(Generic.__module__, sys.modules[Generic.__module__].__file__)
```

```
>>> import torch
>>> import pip._vendor.typing_extensions
>>> import torch._dynamo
typing_extensions .venv/lib/python3.12/site-packages/typing_extensions.py
typing /usr/lib/python3.12/typing.py
typing /usr/lib/python3.12/typing.py
typing /usr/lib/python3.12/typing.py
```

**Comment by hauntsaninja** [comment]
Oh interesting, it's because if you have the real `typing_extensions` and the vendored typing_extensions, then this comparison by identity can hold False https://github.com/python/typing_extensions/blob/196082138aebe36af7f5a1d08691b84dab950eea/src/typing_extensions.py#L3296 because the function is monkeypatched onto here https://github.com/python/typing_extensions/blob/196082138aebe36af7f5a1d08691b84dab950eea/src/typing_extensions.py#L3320 . So if the last `_collect_parameters` to be monkeypatched captures a `NoDefault` that is different from the `NoDefault` set on the ParamSpec you are using https://github.com/python/typing_extensions/blob/196082138aebe36af7f5a1d08691b84dab950eea/src/typing_extensions.py#L1675 , you can have badness!

Minimal repro:
```
import typing_extensions
import pip._vendor.typing_extensions

import typing
P = typing_extensions.ParamSpec("P")
R = typing.TypeVar("R")
class X(typing.Generic[P, R]):
    ...
```

You'll notice that if you swap the import order, it starts working :-)

Upgrading pip works because pip has at least for now seemingly stopped vendoring typing_extensions. Other fixes include not using libraries that use modules vendored by other packages, that's evil. Or if you're really in a pinch, be evil yourself and import pip's internals before anything else executes.
